backend/dal/Services/Concrete/ProjectSnapshotService.cs

```diff
@@ -0,0 +1,196 @@
+using System.Linq;
+using System.Security.Claims;
+
+namespace Pims.Dal.Services
+{
+    /// <summary>
+    /// ProjectSnapshotService class, provides a service layer to interact with project snapshots within the datasource.
+    /// </summary>
+    public class ProjectSnapshotService : BaseService<ProjectSnapshot>, IProjectSnapshotService
+    {
+        #region Variables
+        private readonly PimsOptions _options;
+        #endregion
+
+        #region Constructors
+        /// <summary>
+        /// Creates a new instance of a ProjectSnapshotService, and initializes it with the specified arguments.
+        /// </summary>
+        /// <param name="dbContext"></param>
+        /// <param name="user"></param>
+        /// <param name="service"></param>
+        /// <param name="notifyService"></param>
+        /// <param name="options"></param>
+        /// <param name="logger"></param>
+        public ProjectSnapshotService(PimsContext dbContext, ClaimsPrincipal user, IPimsService service, INotificationService notifyService, IOptions<PimsOptions> options, ILogger<ProjectService> logger) : base(dbContext, user, service, logger)
+        {
+            _options = options.Value;
+        }
+        #endregion
+
+        #region Methods
+        /// <summary>
+        /// Get the total number of project snapshots in the datasource.
+        /// </summary>
+        /// <returns></returns>
+        public int Count()
+        {
+            return this.Context.ProjectSnapshots.Count();
+        }
+
+        /// <summary>
+        /// Get the project snapshot for the specified 'id'.
+        /// </summary>
+        /// <param name="id"></param>
+        /// <exception cref="ArgumentNullException">Argument 'id' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project snapshot does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshot.</exception>
+        /// <returns></returns>
+        public ProjectSnapshot Get(int id)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var snapshot = this.Context.ProjectSnapshots
+                .Include(p => p.FromSnapshot)
+                .FirstOrDefault(p => p.Id == id);
+
+            return snapshot;
+        }
+
+        /// <summary>
+        /// Add the specified project snapshot to the datasource, using the current financial values of the project.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'snapshot' is required.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to add snapshot.</exception>
+        /// <returns></returns>
+        public ProjectSnapshot Add(ProjectSnapshot snapshot)
+        {
+            var originalProject = this.Context.Projects
+                .Include(p => p.Snapshots)
```

Comment:
Do you need to include all the snapshots for the project?  This could kill performance over the years as more and more snapshots are created for projects that sit in the queue.